lib/src/realm_object.dart

```diff
@@ -182,9 +185,18 @@ class RealmCoreAccessor implements RealmAccessor {
         }
         return;
       }
+
       if (value is RealmObject && !value.isManaged) {
         object.realm.add(value, update: update);
       }
+
+      if (propertyMeta.isPrimaryKey && !isInMigration) {
+        final currentValue = realmCore.getProperty(object, propertyMeta.key);
+        if (currentValue != value) {
+          throw RealmException("Primary key cannot be changed (original value: '$currentValue', supplied value: '$value')");
```

Comment:
Do we really need to be so nice. Why not throw the exception on set, as we did previously. Is there a chance this equality compare to not go as planned and skip some values. 
To me we better no allow `set` in this case is just because it adds one more hidden case to think about.  `Set` always throws on PKs is better than `Set` throws on PKs unless the value is the same. This might catch someone off-guard and we better remove all possibilities for mistakes.

Reply:
These are primitive values, so the equality operator should always produce a correct result. I'm generally more inclined to avoid throwing exceptions if absolutely necessary - e.g. if they're using some automapper-like library where you can copy the values from one object to another, it would be kind of annoying if they have to do extra configuration to skip primary key properties.